### ISSUE TITLE
Consistent ordering pmap calls

### ISSUE THREAD

**Issue by sbodenstein** [jax-ml/jax#5391]
When making a call to `pmap`, you can now see which device `id_tap` is [running on](https://github.com/google/jax/pull/5182). However, you still don't have enough information to determine which input batch value this corresponds to, unless you assume consistent ordering (batch slices `(1,2,...,n)` are evaluated by devices with numbers `(1,2,...,n)`). This is true in the current implementation, but not guaranteed in any documentation. Would it make sense to make this a guarantee?

**Comment by gnecula** [comment]
I do not think that we are ready to make this a guarantee, there is still a lot of thinking and movement in the parallelism story. Perhaps a more resilent solution would be to pass to `id_tap` some encoding of the batch slice. I will keep this open as a feature request.

**Comment by vmikulik** [comment]
If I understand correctly, this seems relevant to our usecase. If we want to use `tap_id` to log the intermediate activations in a model that's sharded or replicated via pmap, does this mean that we cannot reconstruct which model shard or which batch element a given activation corresponds to?

**Comment by sbodenstein** [comment]
@gnecula: the use-case of @vmikulik is something logging libraries built on id_tap would like to support (eg. when trying to interpret the activations inside a model, you really do need to know where they came from).

**Comment by gnecula** [comment]
We have deprecated `id_tap` in favor of https://docs.jax.dev/en/latest/external-callbacks.html
